crates/ruff_linter/src/settings/types.rs

```diff
@@ -558,80 +577,231 @@ impl Display for RequiredVersion {
 /// pattern matching.
 pub type IdentifierPattern = glob::Pattern;
 
+/// Like [`PerFile`] but with string globs compiled to [`GlobMatcher`]s for more efficient usage.
 #[derive(Debug, Clone, CacheKey)]
-pub struct CompiledPerFileIgnore {
+pub struct CompiledPerFile<T: CacheKey> {
     pub absolute_matcher: GlobMatcher,
     pub basename_matcher: GlobMatcher,
     pub negated: bool,
-    pub rules: RuleSet,
+    pub data: T,
+}
+
+impl<T: CacheKey> CompiledPerFile<T> {
+    fn new(
+        absolute_matcher: GlobMatcher,
+        basename_matcher: GlobMatcher,
+        negated: bool,
+        data: T,
+    ) -> Self {
+        Self {
+            absolute_matcher,
+            basename_matcher,
+            negated,
+            data,
+        }
+    }
 }
 
-impl Display for CompiledPerFileIgnore {
+impl<T> Display for CompiledPerFile<T>
+where
+    T: Display + CacheKey,
+{
     fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
         display_settings! {
             formatter = f,
             fields = [
                 self.absolute_matcher | globmatcher,
                 self.basename_matcher | globmatcher,
                 self.negated,
-                self.rules,
+                self.data,
             ]
         }
         Ok(())
     }
 }
 
+/// A sequence of [`CompiledPerFile<T>`].
 #[derive(Debug, Clone, CacheKey, Default)]
-pub struct CompiledPerFileIgnoreList {
-    // Ordered as (absolute path matcher, basename matcher, rules)
-    ignores: Vec<CompiledPerFileIgnore>,
+pub struct CompiledPerFileList<T: CacheKey> {
```

Comment:
```suggestion
pub struct CompiledPerFileList<T> {
```

I prefer to avoid having type constraints in struct and instead, only add them on the impls where needed. 

I think the downside of this is that you'd have to implement `CacheKey` manually (because we aren't using Rust edition 2024 yet)

Reply:
Ah that makes sense, I did find it annoying to have to  put `: CacheKey` everywhere. I took a brief look at adding the trait bounds in the derive macro, but I probably need a little more proc macro practice before tackling that!

Reply:
I'm not sure if that's something that you even can do in the proc macro? I'd suggest to simply implement `CacheKey` manually.

Reply:
I think [serde](https://github.com/serde-rs/serde/blob/e3eaa6a3dd6edd701476097182313cdbd73da78c/serde_derive/src/ser.rs#L130) might be able to do it too, again just for my curiositiy, I don't think we need to do this here. `CacheKey` is easy to implement.